### PR TITLE
Merge zn* into z in latency crate

### PR DESCRIPTION
**Pull request by eddie9712**
* Add option "use-expr"
* Add option "declare-publication"
* Remove `zn*` files

### CODE REVIEW

File: latency/src/bin/z_ping.rs [hunk 1]
```diff
@@ -12,112 +12,130 @@
...
     let session = zenoh::open(config).await.unwrap();
     let session = Arc::new(session);
 
+    let mut sub = if opt.use_expr {
+        // declare subscriber
+        let key_expr_pong = session.declare_expr("/test/pong").await.unwrap();
+        session.subscribe(&key_expr_pong).reliable().await.unwrap()
+    } else {
+        session.subscribe("/test/pong").reliable().await.unwrap()
+    };
```
Comment (william-wyliang) [inline latency/src/bin/z_ping.rs:75]: Can this be placed at the original location inside the async block first? Moving it out of the block to remove the barrier synchronization can be done in another commit (to make the code change clearer).

File: latency/src/bin/z_ping.rs [hunk 2]
```diff
@@ -12,112 +12,130 @@
...
+    let mut key_expr_num = 0;
```
Comment (william-wyliang) [inline latency/src/bin/z_ping.rs:76]: Why not use the name "key_expr_ping"?

File: latency/src/bin/z_ping.rs [hunk 3]
```diff
@@ -12,112 +12,130 @@
...
+        key_expr_num = session.declare_expr("/test/ping").await.unwrap();
+        if opt.declare_publication {
+            session.declare_publication(key_expr_num).await.unwrap();
+        }
+    } else {
+        if opt.declare_publication {
+            session.declare_publication("/test/ping").await.unwrap();
+        }
+    }
+
```
Comment (william-wyliang) [inline latency/src/bin/z_ping.rs:87]: Consider to put this declaration close to the place it's used below, before or in the loop construct.

File: latency/src/bin/z_ping.rs [hunk 4]
```diff
@@ -12,112 +12,130 @@
...
-
         while let Some(sample) = sub.next().await {
-                  let mut payload_reader = sample.value.payload.reader();   
-                  let mut count_bytes = [0u8; 8];
-                  if payload_reader.read_exact(&mut count_bytes){
-                    let count = u64::from_le_bytes(count_bytes);
+            let mut payload_reader = sample.value.payload.reader();
+            let mut count_bytes = [0u8; 8];
+            if payload_reader.read_exact(&mut count_bytes) {
+                let count = u64::from_le_bytes(count_bytes);
```
Comment (william-wyliang) [inline latency/src/bin/z_ping.rs:100]: Consider to have the `rustfmt` done in an independent commit.

File: latency/src/bin/z_ping.rs [hunk 5]
```diff
@@ -12,112 +12,130 @@
...
+                .congestion_control(CongestionControl::Block)
+                .await
+                .unwrap();
+        } else {
+            session
+                .put("/test/ping", payload)
+                .congestion_control(CongestionControl::Block)
+                .await
+                .unwrap();
+        }
```
Comment (william-wyliang) [inline latency/src/bin/z_ping.rs:138]: A suggestion here. By way of declare `writer` first, this long "copy-paste" style code can be removed. Refer to the usage [here](https://github.com/william-wyliang/zenoh-perf/blob/665cd2bc32b74d3d8ce0a11c78a9fc54bd40181e/throughput/src/bin/z_put_thr.rs#L125).

File: latency/src/bin/z_ping.rs [hunk 6]
```diff
@@ -144,33 +175,42 @@ async fn single(opt: Opt, config: Config) {
         payload[0..8].copy_from_slice(&count_bytes);
 
         let now = Instant::now();
-        session
-            .put("/test/ping", payload)
```
Comment (william-wyliang) [inline latency/src/bin/z_ping.rs:148]: As above, consider to declare `writer` here to simply the code here and below.

File: latency/src/bin/z_ping.rs [hunk 7]
```diff
@@ -131,11 +149,24 @@ async fn single(opt: Opt, config: Config) {
...
+    if opt.use_expr {
+        key_expr_num = session.declare_expr("/test/ping").await.unwrap();
+        if opt.declare_publication {
+            session.declare_publication(key_expr_num).await.unwrap();
+        }
+    } else {
+        if opt.declare_publication {
+            session.declare_publication("/test/ping").await.unwrap();
+        }
+    }
```
Comment (william-wyliang) [inline latency/src/bin/z_ping.rs:169]: Similar concern as that in `parallel()`.

File: latency/src/bin/z_pong.rs [hunk 8]
```diff
@@ -57,18 +65,40 @@ async fn main() {
...
+    if opt.use_expr {
+        key_expr_num = session.declare_expr("/test/pong").await.unwrap();
+        if opt.declare_publication {
+            session.declare_publication(key_expr_num).await.unwrap();
+        }
+    } else {
+        if opt.declare_publication {
+            session.declare_publication("/test/pong").await.unwrap();
+        }
+    }
```
Comment (william-wyliang) [inline latency/src/bin/z_pong.rs:86]: Similar concerns as those commented for `z_ping`.

**Review by william-wyliang — CHANGES_REQUESTED** [review]
- Consider to split into 2 or 3 commits for different changes for moving codes, adding new flags and the rustfmt operation.
- Consider ways to reduce "copy-paste" style codes.

**Comment by william-wyliang** [comment]
By the way, a good habit or practice is to add verification method (and/or checklist) in your PRs. FYR.

File: latency/src/bin/z_ping.rs (hunk 4 above)
Comment (eddie9712) [reply latency/src/bin/z_ping.rs:100]: Ok.

File: latency/src/bin/z_ping.rs (hunk 5 above)
Comment (eddie9712) [reply latency/src/bin/z_ping.rs:138]: IMO, the `z_ping` case is differ to the case of `z_put_thr`, because in `z_ping` case, we need to create the payload in each iteration,  hence we need to create the writer inside the loop ([here](https://github.com/william-wyliang/zenoh-perf/blob/665cd2bc32b74d3d8ce0a11c78a9fc54bd40181e/throughput/src/bin/z_put_thr.rs#L98)), I am not sure is it a good idea to put these actions into the loop (includes `declare_publication` and `declare_expr`).

File: latency/src/bin/z_ping.rs (hunk 1 above)
Comment (eddie9712) [reply latency/src/bin/z_ping.rs:75]: Ok, I will split it into different commits.

File: latency/src/bin/z_ping.rs (hunk 2 above)
Comment (eddie9712) [reply latency/src/bin/z_ping.rs:76]: Originally, I just want to separate two different possible "key expression" (might be string or numerical Id), what do you think?

Comment (william-wyliang) [reply latency/src/bin/z_ping.rs:76]: I think source code readability is quit important. It will be good to name it in a more meaningful way.

File: latency/src/bin/z_ping.rs (hunk 5 above)
Comment (william-wyliang) [reply latency/src/bin/z_ping.rs:138]: `session.put()` already creates a writer, it's just one way to avoid duplicating the whole builder statement. You can consider that and make your own judgement based on the complexity when `declare_publication` and `declare_expr` are also included.
